src/logic/events/WorkspaceEvents.swift

```diff
@@ -14,11 +14,30 @@ class WorkspaceEvents {
         let diff = Array(workspaceApps.symmetricDifference(previousValueOfRunningApps))
         if change.kind == .insertion {
             debugPrint("OS event", "apps launched", diff.map { ($0.processIdentifier, $0.bundleIdentifier) })
-            Applications.addRunningApplications(diff)
         } else if change.kind == .removal {
             debugPrint("OS event", "apps quit", diff.map { ($0.processIdentifier, $0.bundleIdentifier) })
             Applications.removeRunningApplications(diff)
+            previousValueOfRunningApps = workspaceApps
+        }
+    }
+
+    static func registerFrontAppChangeNote() {
+        NSWorkspace.shared.notificationCenter.addObserver(self, selector: #selector(receiveFrontAppChangeNote(_:)), name: NSWorkspace.didActivateApplicationNotification, object: nil)
+    }
+
+    // We add apps when we receive a didActivateApplicationNotification notification, not when we receive an apps launched, because any app will have an apps launched notification.
+    // But we are only interested in apps that have windows. We think that since an app can be activated, it must have a window, and subscribing to its window event makes sense and is likely to work, even if it requires multiple retries to subscribe.
+    // I'm not very sure if there is an edge case, but so far testing down the line has not revealed it.
+    // When we receive the didActivateApplicationNotification notification, NSRunningApplication.isActive=true, even if the app is not the frontmost window anymore.
+    // If we go to add the application when we receive the message of apps launched, at this time NSRunningApplication.isActive may be false, and try axUiElement.windows() may also throw an exception.
+    // For those background applications, we don't receive notifications of didActivateApplicationNotification until they have their own window. For example, those menu bar applications.
```

Comment:
If this is true, then your approach here to listen to `.didActivateApplicationNotification` is genius! Did you think or it yourself or was it used by other projects?

Also, if I summarize this change, it seems you found an API that gives us properly filtered apps (only apps with windows), but that's only for apps launched *after* AltTab is started. For apps launched before AltTab is started, we still have to filter out processes with `activationPolicy`, `XPC`, etc. But luckily, apps launched before AltTab should not throw an exception on `.windows()`.

And then you mention the edge-case of an app launching at the same time as AltTab, which would be annoying. This is actually a very important edge-case, unfortunately. That's what happens when you log into macOS. All the startup apps launch at once, concurrently. This means AltTab is launched, and all sorts of app at the very same time.

I guess for that scenario, `didActivateApplicationNotification` will not help us, and we will need to rely on the other existing checks (XPC, etc), and with the new `250ms` delay, it may be an acceptable situation?

Am I understanding correctly the whole approach?

Reply:
> Did you think or it yourself or was it used by other projects?

At first I was thinking that when we open an application, as long as they show a window, then a change in window switching is bound to happen, and if we can find a way to observe this change, we will be able to tell that the program is equipped with a window. Then I used some window management software, such as automatic window layout APP, like AppGrid, Rectangle, and found that they can observe this change very well, so I went to study their code, and then found in Rectangle's code that it subscribes to `.didActivateApplicationNotification` notifications. I was curious, so I wrote some test code to see the effect. Turns out it worked and was able to solve half of the problem, the other half still had to be figured out.

> And then you mention the edge-case of an app launching at the same time as AltTab, which would be annoying. This is actually a very important edge-case, unfortunately. That's what happens when you log into macOS. All the startup apps launch at once, concurrently. This means AltTab is launched, and all sorts of app at the very same time.

That's my concern, and I still need to do some testing on this situation.

> I guess for that scenario, didActivateApplicationNotification will not help us, and we will need to rely on the other existing checks (XPC, etc), and with the new 250ms delay, it may be an acceptable situation?

This edge case only happens when `wasLaunchedBeforeAltTab=true`, at which point we go to the `axUiElement.windows()` test. If it happens that the program takes a long time to load, then `axUiElement.windows()` will most likely throw an exception, and at that point we do need to detect it by other means (e.g. XPC, etc), otherwise we will lose the window. Of course, reopening the program that lost the window, AltTab will capture its window without any problem.

Reply:
For that edge-case, we probably don't need to restart the applications that are missing windows, we can do a check when we get the `.didActivateApplicationNotification` notification, and if the application is not in the AltTab list of existing windows, then we add it.